Add forward AD formulas for {batch,layer,group}_norm

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71026
* #70468
* __->__ #70355

Differential Revision: [D33405362](https://our.internmc.facebook.com/intern/diff/D33405362)

soulitzer committed:
Add forward AD formulas for some {batch,layer,group}_norm

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"




[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4546,6 +4546,147 @@ Tensor cumprod_jvp(Tensor self_t, Tensor self_p, Tensor result, int dim) {
+  for (const auto dim : c10::irange(view_size.size())) {
+    if (dim != 1) {
+      numel *= input_t.size(dim);
+      view_size[dim] = 1;
```
Is this modifying the sizes_ attribute of `input_t`? Or is the `.vec()` creating a new storage for the vector out of place?

soulitzer replied:
Yes I think it is out of place. It should be (5) from https://en.cppreference.com/w/cpp/container/vector/vector

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4546,6 +4546,147 @@ Tensor cumprod_jvp(Tensor self_t, Tensor self_p, Tensor result, int dim) {
+  auto view_size = input_t.sizes().vec();
+  int64_t numel = 1;
+  for (const auto dim : c10::irange(view_size.size())) {
+    if (dim != 1) {
```
batchnorm doesn't support no-batch dim?

soulitzer replied:
Looking at https://github.com/pytorch/pytorch/issues/60585
- yes, batch norm semantically is incompatible with no-batch dim
- for group norm it is BC-breaking, so a potential plan seems to be to create a new  `nn.Module` entirely that would handle the no batch dim case
- layer norm should be fine as long as we handle the case where normalized shape has the same dimension as input. I do see a test case in OpInfo handling this so we should be ok

cc @jbschlosser

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4546,6 +4546,147 @@ Tensor cumprod_jvp(Tensor self_t, Tensor self_p, Tensor result, int dim) {
+  if (train) {
+    mean_p = saved_mean.view(view_size);
+    invstd_p = saved_invstd.view(view_size);
+    auto invstd_t = -invstd_p.pow(3) * ((input_t - input_t.mean(dims, true)) * (input_p - mean_p)).sum(dims, true) / numel;
```
There are quite a few things that could be done inplace here.
I think we want to do these. Keeping in mind the "composite compliance"

cc @zou3519 is there a doc/wiki/inline comment I can link here that states the rules of what is allowed or not?

zou3519 replied:
This: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md#compositeimplicitautograd-compliance , but it doesn't include the information about the in-place yet

zou3519 replied:
Try this as well: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorSubclassLikeUtils.h

albanD replied:
Thanks!
Should we have a wiki or doc note that points to these?

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4546,6 +4546,147 @@ Tensor cumprod_jvp(Tensor self_t, Tensor self_p, Tensor result, int dim) {
+  } else {
+    TORCH_INTERNAL_ASSERT(
+        running_mean.has_value() && running_var.has_value(),
+        "Expect running_mean and running_var to have value when train=true");
```
train=false

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4546,6 +4546,147 @@ Tensor cumprod_jvp(Tensor self_t, Tensor self_p, Tensor result, int dim) {
+    } else {
+      numel *= input_t.size(i);
+      view_size[i] = 1;
+      dims[i - view_size.size() + normalized_shape.size()] = i;
```
Maybe you want to use `push_back()` here to make this logic simpler?

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4546,6 +4546,147 @@ Tensor cumprod_jvp(Tensor self_t, Tensor self_p, Tensor result, int dim) {
+  int64_t N = input_p.size(0);
+  int64_t C = input_p.size(1);
+
+  auto input_t_reshaped = input_t.view({1, N * groups, N ? -1 : 1});
```
if 1 is valid, then -1 is also always valid no?

soulitzer replied:
If N is 0, -1 is not allowed:

`RuntimeError: cannot reshape tensor of 0 elements into shape [0, -1] because the unspecified dimension size -1 can be any value and is ambiguous`

albanD replied:
Ho interesting. Sounds good!

albanD reviewed (commented):
Looks good.
There are some small optimizations we want to do though.

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"




[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405362).

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"


Differential Revision: [D33405362](https://our.internmc.facebook.com/intern/diff/D33405362)

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"


Differential Revision: [D33405362](https://our.internmc.facebook.com/intern/diff/D33405362)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405362).

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"


Differential Revision: [D33405362](https://our.internmc.facebook.com/intern/diff/D33405362)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405362).

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"


Differential Revision: [D33405362](https://our.internmc.facebook.com/intern/diff/D33405362)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405362).

albanD reviewed (approved):
Update looks good!

soulitzer committed:
Update on "Add forward AD formulas for {batch,layer,group}_norm"


Differential Revision: [D33405362](https://our.internmc.facebook.com/intern/diff/D33405362)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33405362).
